### PR TITLE
Implement delta backups for Cassandra, part 1

### PR DESCRIPTION
**Pull request by dmitry-potepalov**
Add the concept of a lighter "delta" backup to Astacus. Lighter "delta" backups contain only the data reported as such by the database itself (e.g. via commit logs or Cassandra incremental backups). Compared to the "regular" snapshot-based backups, delta backups:
* do not require listing the hexdigest storage to create the backup
* do not require to build a full list of files present in the database to determine the diff
* can be restored one on top of the other (during node replacement this allows to do "catch-up" on the replacement before draining&removing the old node, decreasing the window of time the cluster is running with RF-1 nodes)

The delta APIs re-use the download and snapshot methods already present, but:
* deltas are gathered and uploaded using a different Snapshotter instance targeted at the delta globs
* Cassandra incremental backups require the user to remove the incremental files once the user's done with them, so an additional Clear step is added at the end

This PR is also an invitation to a discussion: do we need a separate "delta backup" concept in Astacus? My main "pro" arguments are:
* enables the "catch-up" behavior during replacement
* looks more natural in a PITR context (technically one could increase backup frequency to "each 5 minutes" and include the commitlogs since the latest full backup to allow restoring to "somewhere in between", but that looks a bit counterintuitive: why would a snapshot include something that depends on a previous snapshot?)
* doesn't require to iterate through all the files (and hexdigests) every time we want to do a backup (might not be compelling enough on it's own, but it's a nice bonus)

### CODE REVIEW

**Comment by joelynch** [comment]
So is the idea that you would say once a day do a "full" backup (which is like the checkpoint) and then very frequently do the delta backups and then when restoring download the latest full backup and all the delta backups that occurred after that time?

I think this is great, makes sense to me.  The delta backups are quite different and I think trying to shoehorn them into the existing backups is probably not a good idea.

File: astacus/client.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def _run_op(op, args, *, json=None, data=None) -> bool:
 
 
 def _run_backup(args) -> bool:
-    return _run_op("backup", args)
+    return _run_op("deltabackup" if args.delta else "backup", args)
```
Comment (kmichel-aiven) [inline astacus/client.py:54]: They are different enough that I think it would be better to have a separate command for it instead of an option. (For clarity, to keep the parallel with the coordinator API, and because if we add options to only backup or only delta backups, the argparse and help will be simpler and avoid "this option only applies if you also enable this other option").

File: astacus/client.py [hunk 2]
```diff
@@ -143,7 +143,8 @@ def _run_list(args) -> bool:
     storage_name = ""
     if args.storage:
         storage_name = f"?storage={args.storage}"
-    r = http_request(f"{args.url}/list{storage_name}", caller="client._run_list")
+    method = "delta/list" if args.delta else "list"
```
Comment (kmichel-aiven) [inline astacus/client.py:146]: A bit confusing to say "method" for a path in an HTTP context :)

File: astacus/coordinator/api.py [hunk 3]
```diff
@@ -81,6 +82,12 @@ async def backup(*, c: Coordinator = Depends(), op: BackupOp = Depends(BackupOp.
     return c.start_op(op_name=OpName.backup, op=op, fun=runner)
 
 
+@router.post("/deltabackup")
```
Comment (kmichel-aiven) [inline astacus/coordinator/api.py:85]: For consistency with /delta/list and the same pattern used in node ops, this could be better as `/delta/backup`

File: astacus/coordinator/api.py [hunk 4]
```diff
@@ -30,6 +30,7 @@ class OpName(StrEnum):
     """(Long-running) operations defined in this API (for coordinator)"""
 
     backup = "backup"
+    delta_backup = "deltabackup"
```
Comment (kmichel-aiven) [inline astacus/coordinator/api.py:33]: Can this be kept consistent with the name and use `delta_backup` for the value or do we need the non-underscore ?

File: astacus/node/api.py [hunk 5]
```diff
@@ -115,11 +130,26 @@ def upload_result(*, op_id: int, n: Node = Depends()):
...
+    op, _ = n.get_op_and_op_info(op_id=op_id, op_name=OpName.upload)
+    return op.result
+
+
 @router.post("/download")
 def download(req: ipc.SnapshotDownloadRequest, n: Node = Depends()):
     if not n.state.is_locked:
         raise HTTPException(status_code=409, detail="Not locked")
-    return DownloadOp(n=n, op_id=n.allocate_op_id(), stats=n.stats, req=req).start()
+    return DownloadOp(n=n, op_id=n.allocate_op_id(), stats=n.stats, req=req).start(n.get_or_create_snapshotter)
```
Comment (kmichel-aiven) [inline astacus/node/api.py:152]: This looks a bit too hard to follow with the init parameters but also a callback passed later, the init takes a `req` and then the callback will be called with things from `req` that was passed just before that. I think it's worth simplifying and just passing a configured snapshotter from the API to the Op, without callbacks.

**Review by kmichel-aiven — CHANGES_REQUESTED** [review]
I'm also not sure the multiple snapshotter in the state and the clear operations align well with https://github.com/Aiven-Open/astacus/pull/139

File: astacus/coordinator/api.py (hunk 3 above)
Comment (dmitry-potepalov) [reply astacus/coordinator/api.py]: the routing and result routing was built with the assumption that the operation name is a single word. I cheated a bit and reused the "backup" op type, should be good enough.

**Comment by dmitry-potepalov** [comment]
> I'm also not sure the multiple snapshotter in the state and the clear operations align well with https://github.com/Aiven-Open/astacus/pull/139 

took a look, I don't see any major issues, just introduce a delta Snapshot under a different key in app state and it should work. Could you maybe pinpoint some specific issue?

File: astacus/coordinator/plugins/base.py [hunk 6]
```diff
@@ -152,10 +153,12 @@ class UploadBlocksStep(Step[List[ipc.SnapshotUploadResult]]):
 
     storage_name: str
     validate_file_hashes: bool = True
+    upload_request: str = "upload"
+    list_hexdigests_step: type = ListHexdigestsStep
```
Comment (kmichel-aiven) [inline astacus/coordinator/plugins/base.py:157]: The type is too vague here (and I'm surprised mypy doesn't complain), it should be `type[Step[Set[str]]]`, this allows static validatoin that the return value of `context.get_result` has the correct type.




Ah, I know why mypy didn't complain :( 
```python
def build_node_index_datas(
    *, hexdigests, snapshots: List[ipc.SnapshotResult], node_indices: List[int]
) -> List[NodeIndexData]:
```

File: astacus/coordinator/plugins/base.py [hunk 7]
```diff
@@ -164,7 +167,28 @@ async def run_step(self, cluster: Cluster, context: StepsContext) -> List[ipc.Sn
...
             node_index_datas,
             validate_file_hashes=self.validate_file_hashes,
+            upload_request=self.upload_request,
+        )
+
+
+@dataclasses.dataclass
+class SnapshotClearStep(Step[List[ipc.NodeResult]]):
+    """
+    Request to clear the source hierarcy of the snapshotter on all nodes.
```
Comment (kmichel-aiven) [inline astacus/coordinator/plugins/base.py:177]: nit. hierarchy

File: astacus/node/api.py [hunk 8]
```diff
@@ -173,3 +223,38 @@ def cassandra(req: Union[ipc.NodeRequest, ipc.CassandraStartRequest], subop: ipc
...
+            for group in req.groups
+            if not any(existing_group.root_glob == group.root_glob for existing_group in groups)
+        ]
+    return groups
+
+
+def snapshotter_from_snapshot_req(req: SnapshotReq, n: Node) -> Snapshotter:
+    groups = groups_from_snapshot_req(req)
+    snapshotter = n.get_or_create_snapshotter(groups)
+    assert snapshotter is not None
```
Comment (kmichel-aiven) [inline astacus/node/api.py:252]: Why do we need this assert ?

File: astacus/node/api.py [hunk 9]
```diff
@@ -173,3 +223,38 @@ def cassandra(req: Union[ipc.NodeRequest, ipc.CassandraStartRequest], subop: ipc
...
+    return snapshotter
+
+
+def delta_snapshotter_from_snapshot_req(req: SnapshotReq, n: Node) -> Snapshotter:
+    groups = groups_from_snapshot_req(req)
+    snapshotter = n.get_or_create_delta_snapshotter(groups)
+    assert snapshotter is not None
```
Comment (kmichel-aiven) [inline astacus/node/api.py:259]: Same here, why do we need this assert ?

**Review by kmichel-aiven — DISMISSED** [review]

File: astacus/coordinator/api.py [hunk 10]
```diff
@@ -114,22 +120,30 @@ async def _list_backups(*, req: ipc.ListRequest = ipc.ListRequest(), c: Coordina
...
+
+
 @router.post("/cleanup")
 async def cleanup(*, op: CleanupOp = Depends(CleanupOp.create), c: Coordinator = Depends()):
     runner = await op.acquire_cluster_lock()
     return c.start_op(op_name=OpName.cleanup, op=op, fun=runner)
 
 
 @router.get("/{op_name}/{op_id}")
+@router.get("/delta/{op_name}/{op_id}")
```
[the remark this answers is not in the document]
Comment (dmitry-potepalov) [reply astacus/coordinator/api.py:136]: because operation status url depends on the request path, therefore status urls for `/delta/X` look like `/delta/X/N`
```
    def start_op(self, *, op: Op, op_name: str, fun: Callable[[], Any]) -> Op.StartResult:
        ...
        url = self.request_url
        status_url = urlunsplit((url.scheme, url.netloc, f"{url.path}/{op.op_id}", "", ""))
```

File: astacus/coordinator/plugins/base.py [hunk 11]
```diff
@@ -99,6 +97,7 @@ class SnapshotStep(Step[List[ipc.SnapshotResult]]):
     """
 
     snapshot_groups: Sequence[SnapshotGroup]
+    snapshot_request: str = "snapshot"
```
[the remark this answers is not in the document]
Comment (dmitry-potepalov) [reply astacus/coordinator/plugins/base.py:100]: but this is `SnapshotStep`, not `SnapshotRequestV2`; steps are internal to the coordinator and not serialized/sent between nodes. The `snapshot_request` parameter of the step is either `snapshot` or `delta/snapshot` depending on the context.

[the remark this answers is not in the document]
Comment (kmichel-aiven) [reply astacus/coordinator/plugins/base.py:100]: :facepalm: I misread the diff...

File: astacus/node/api.py (hunk 8 above)
Comment (dmitry-potepalov) [reply astacus/node/api.py]: hmm, indeed seems that we don't; probably repeated the one from inside the op

**Review by kmichel-aiven — APPROVED** [review]
